xcm/pallet-xcm/src/lib.rs

```diff
@@ -1203,8 +1203,10 @@ impl<T: Config> Pallet<T> {
 			BuyExecution { fees, weight_limit },
 			DepositAsset { assets: Wild(AllCounted(max_assets)), beneficiary },
 		]);
-		let mut message =
-			Xcm(vec![WithdrawAsset(assets), InitiateTeleport { assets: Wild(All), dest, xcm }]);
+		let mut message = Xcm(vec![
+			WithdrawAsset(assets),
+			InitiateTeleport { assets: Wild(AllCounted(max_assets)), dest, xcm },
```

Comment:
I guess, we need to align and think about more things:

weights for call `limited_teleport_assets` [here](https://github.com/paritytech/polkadot/pull/6980/files#diff-8a9c007ce54da1bb40d9a7f6c8c796d7f07b38d7a4129dca6c015fb67987abefR1071):
```
assets: Wild(All)
```

weights for call `teleport_assets` [here](https://github.com/paritytech/polkadot/pull/6980/files#diff-8a9c007ce54da1bb40d9a7f6c8c796d7f07b38d7a4129dca6c015fb67987abefR803):
```
assets: Wild(AllCounted(count)),
```

`do_teleport_assets` is called from both extrinsics
max_assets is `const MAX_ASSETS_FOR_TRANSFER: usize = 2;`

---

`remote_weight` guessing is using `max_assets` as worst case, ok, 
but for the real message why not to use exact count?
```
Wild(AllCounted(assets.len())
``

in cases where fee is paid by the same asset (DOTs/KSMs) it will be cheaper then `Wild(AllCounted(max_assets))`

Reply:
As far as I know `max_assets` in this case is `assets.len()`.
`MAX_ASSETS_FOR_TRANSFER` is used in an 
`ensure!(assets.len() <= MAX_ASSETS_FOR_TRANSFER, Error::<T>::TooManyAssets);` in do_teleport_assets. So in this case this is correct, is it not?

Reply:
https://github.com/paritytech/polkadot/blob/21ee3cd452d297fae36fe4ea2d0a8c33ef6e156b/xcm/pallet-xcm/src/lib.rs#L1173 - here's the ensure.

https://github.com/paritytech/polkadot/blob/21ee3cd452d297fae36fe4ea2d0a8c33ef6e156b/xcm/pallet-xcm/src/lib.rs#L1184 - `let max_assets = assets.len()`

Reply:
The only thing we could do here is perhaps rename `max_assets` to `total_assets` to avoid this kind of confusion going forward.

Reply:
ah, yeah, it is ok, sorry for confusion,
my brain thought that it is:
`let max_assets = MAX_ASSETS_FOR_TRANSFER`
but it is
`let max_assets = assets.len() as u32;`

`max_assets` is kind of misleading, maybe `assets_count`